src/ast/passes/codegen_llvm.cpp

```diff
@@ -2553,7 +2555,20 @@ void CodegenLLVM::visit(Subprog &subprog)
                                               0);
 
   Function *func = Function::Create(
-      func_type, Function::InternalLinkage, subprog.name(), module_.get());
+      func_type, Function::ExternalLinkage, subprog.name(), module_.get());
+
+  // Mark function as NoInline. This has several reasons:
+  // - inlining takes up BPF instructions
+  // - LLVM sometimes inlines recursive calls into loops, which can be then
+  //   rejected by the BPF verifier for code where a recursive call would
+  //   not
+  AttributeList attrs;
+  attrs = attrs.addFnAttribute(b_.getContext(), Attribute::NoInline);
```

Comment:
Looks like this doesn't exist for llvm 12 and 13 (I think `addFnAttribute` was added in 14) so we might need a `#if LLVM_VERSION_MAJOR`